### PR TITLE
 adding Nomad scheduler support for Heron

### PR DESCRIPTION
**Pull request by jerrypeng**
This PR contains the core code for running Heron on Nomad.

In a subsequent PR, I will add additional scripts to allow users to easily deploy a standalone cluster running Heron.

Something like this:

heron standalone set masters

heron standalone set slaves

heron standalone start cluster

### CODE REVIEW

**Comment by jerrypeng** [comment]
@srkukarni @maosongfu @nwangtw

File: heron/executor/src/python/heron_executor.py [hunk 1]
```diff
@@ -978,7 +982,9 @@ def start_state_manager_watches(self):
     Receive updates to the packing plan from the statemgrs and update processes as needed.
     """
     statemgr_config = StateMgrConfig()
-    statemgr_config.set_state_locations(configloader.load_state_manager_locations(self.cluster))
+    statemgr_config.set_state_locations(configloader.load_state_manager_locations(
```
Comment (srkukarni) [inline heron/executor/src/python/heron_executor.py]: can you comment on what's going on here?

Comment (jerrypeng) [reply heron/executor/src/python/heron_executor.py]: I add this because specifying an override via commandline for the config "heron.statemgr.connection.string" does not get propagated into the heron_executor.  Thus, if I want to specify the location of zookeeper via command line, I cannot.

File: heron/config/src/yaml/BUILD [hunk 2]
```diff
@@ -14,7 +14,7 @@ filegroup(
 
 filegroup(
     name = "conf-yaml",
-    srcs = glob(["conf/**/*.yaml"] + ["conf/**/*.aurora"]),
+    srcs = glob(["conf/**/*.yaml"] + ["conf/**/*.aurora"] + ["conf/**/*.sh"]),
```
Comment (nwangtw) [inline heron/config/src/yaml/BUILD]: Hmm. Is it possible to avoid the .sh files here?

File: heron/downloaders/src/java/com/twitter/heron/downloader/DownloadRunner.java [hunk 3]
```diff
@@ -34,6 +34,7 @@ public static void main(String[] args) throws Exception {
     final Path topologyDestination = Paths.get(destination);
 
     final File file = topologyDestination.toFile();
+
```
Comment (nwangtw) [inline heron/downloaders/src/java/com/twitter/heron/downloader/DownloadRunner.java]: revert this change

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+package com.twitter.heron.downloader;
+
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Path;
+
+public class FileDownloader implements Downloader {
+  @Override
+  public void download(URI uri, Path destination) throws Exception {
+    final URL url = uri.toURL();
```
Comment (nwangtw) [inline heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: do we need try/cache here?

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+package com.twitter.heron.downloader;
+
+import java.net.URI;
+import java.net.URL;
+import java.nio.file.Path;
+
+public class FileDownloader implements Downloader {
```
Comment (nwangtw) [inline heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: Why a new downloader is added? HttpDownloader is not enough in this case? The implementation seems to be the same.

If there is a new case, please add comment to explain. My concern is that in future people dont know which one to use or could mis-use them.

File: heron/schedulers/src/java/BUILD [hunk 6]
```diff
@@ -53,6 +53,25 @@ kubernetes_deps_files = \
...
+          "@commons_codec_commons_codec//jar",
+          "@commons_logging_commons_logging//jar",
+          "@org_apache_httpcomponents_http_client//jar",
+          "@org_apache_httpcomponents_http_core//jar",
+          "@org_bouncycastle_bcpkix_jdk15on//jar",
+          "@org_bouncycastle_bcprov_jdk15on//jar",
+      ]
+    
+nomad_deps_files = \
+    scheduler_deps_files + nomad_sdk_deps +[
```
Comment (nwangtw) [inline heron/schedulers/src/java/BUILD]: space between "+" and "["

File: heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/AppsV1beta1Controller.java [hunk 7]
```diff
@@ -228,7 +228,7 @@ boolean isStatefulSet() {
     final Config runtimeConfiguration = getRuntimeConfiguration();
     final String[] executorCommand =
         SchedulerUtils.getExecutorCommand(configuration, runtimeConfiguration,
-            containerId, ports);
+            Integer.valueOf(containerId), ports);
```
Comment (cckellogg) [inline heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/AppsV1beta1Controller.java]: This needs to be a string because we use an environment variable for the container id

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java [hunk 8]
```diff
@@ -189,29 +189,10 @@ private SchedulerUtils() {
...
-   * Utils method to construct the command to start heron-executor
-   *
-   * @param config The static config
-   * @param runtime The runtime config
-   * @param shardId the executor/container index
-   * @param ports a map of ports to use where the key indicate the port type and the
-   * value is the port
-   * @return String[] representing the command to start heron-executor
-   */
-  public static String[] getExecutorCommand(
```
Comment (cckellogg) [inline heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: Kubernetes needs to be able to send a string as the shard id

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java (hunk 5 above)
Comment (jerrypeng) [reply heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: So I added this because the heron_downloader couldn't handle URIs like: "file:///foo/bar".  This is the same implementation as the HttpDownloader, but I am not sure if we should point URIs with a prefix of "file" to the the http downloader or just have a new file downloader.  The reason to why we need to support copy of files by the heron_downloader is because there could be instances where the user is distributing JARs via some sort of network storage and we need to copy and extract the jars from the network storage location into the local heron directory structure

File: heron/schedulers/src/java/BUILD (hunk 6 above)
Comment (jerrypeng) [reply heron/schedulers/src/java/BUILD]: will add

File: heron/downloaders/src/java/com/twitter/heron/downloader/DownloadRunner.java (hunk 3 above)
Comment (jerrypeng) [reply heron/downloaders/src/java/com/twitter/heron/downloader/DownloadRunner.java]: will revert

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java (hunk 4 above)
Comment (jerrypeng) [reply heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: though this is really done once on the start of the topology

File: heron/config/src/yaml/BUILD (hunk 2 above)
Comment (jerrypeng) [reply heron/config/src/yaml/BUILD]: We need the heron_nomad.sh script to start heron_executors in nomad.  Not sure where else would be more appropriate to put this since the script needs to be available on where you install the heron cli.  The heron.aurora script is placed in the aurora config folder.

File: heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java (hunk 8 above)
Comment (jerrypeng) [reply heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java]: will change back

File: heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/AppsV1beta1Controller.java (hunk 7 above)
Comment (jerrypeng) [reply heron/schedulers/src/java/com/twitter/heron/scheduler/kubernetes/AppsV1beta1Controller.java]: will change back

**Comment by jerrypeng** [comment]
@srkukarni @nwangtw @cckellogg thanks for your reviews!  I have addressed your comments.

File: heron/config/src/yaml/conf/nomad/statemgr.yaml [hunk 9]
```diff
@@ -0,0 +1,23 @@
+# local state manager class for managing state in a persistent fashion
+heron.class.state.manager: com.twitter.heron.statemgr.zookeeper.curator.CuratorStateManager
+
+# local state manager connection string
+heron.statemgr.connection.string:  127.0.0.1:2181
```
Comment (nwangtw) [inline heron/config/src/yaml/conf/nomad/statemgr.yaml]: why 2181? I think it is the default port for zk so maybe better to avoid if possible.

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java (hunk 4 above)
Comment (nwangtw) [reply heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: kk.

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java (hunk 5 above)
Comment (nwangtw) [reply heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: Got it. thx. Please add a comment about the use case. An example could be good enough.

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  @After
+  public void after() throws Exception {
+    scheduler.close();
+  }
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+
```
Comment (nwangtw) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: Remove the empty line.

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+  }
+
+  @AfterClass
+  public static void afterClass() throws Exception {
+
```
Comment (nwangtw) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: Remove the empty line.

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+    List<Job> jobList = Arrays.asList(jobs);
+
+    // Fail to restart
+    PowerMockito.when(NomadScheduler.getTopologyJobs(Mockito.any(NomadApiClient.class),
+        Mockito.anyString())).thenReturn(jobList);
+    PowerMockito.doThrow(new RuntimeException()).when(NomadScheduler.class,
+        "restartJobs", Mockito.any(), anyVararg());
+    Assert.assertFalse(scheduler.onRestart(restartTopologyRequest));
+
+    // succeed to restart
```
Comment (nwangtw) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: nit: s/succeed/Succeed

There are a few of them in this file

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+        = Scheduler.KillTopologyRequest.newBuilder()
+        .setTopologyName(TOPOLOGY_NAME)
+        .build();
+
+    scheduler.initialize(this.mockConfig, this.mockRuntime);
+
+    Job[] jobs = {new Job(), new Job(), new Job()};
+    List<Job> jobList = Arrays.asList(jobs);
+
+    // fail to kill
```
Comment (nwangtw) [inline heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: nit: s/fail/Fail

File: heron/schedulers/tests/java/BUILD [hunk 14]
```diff
@@ -68,6 +68,26 @@ kubernetes_deps_files = [
     "//third_party/java:kubernetes-java-client",
 ]
 
+nomad_sdk_deps = [
+          "@com_hashicorp_nomad//jar",
```
Comment (nwangtw) [inline heron/schedulers/tests/java/BUILD]: nit: bad indent.

File: heron/config/src/yaml/BUILD (hunk 2 above)
Comment (nwangtw) [reply heron/config/src/yaml/BUILD]: heron.aurora might be ok (maybe not ideal though) because aurora file is a config and used by aurora cli.

.sh file is worse than aurora file. :(  So I am wondering if it is possible to avoid this .sh file and have a config to be used by nomad CLI?

I don't know nomad enough and I am open to other thoughts/suggestions. Also this is not a blocker if this is not feasible.

Comment (jerrypeng) [reply heron/config/src/yaml/BUILD]: The .sh file is actually read by the scheduler and submitted to nomad as part of a job description in json format.  I can inline the contents of the .sh script in the scheduler code, but that might not be as clear to the user as just having a file with comments in it

File: heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java (hunk 5 above)
Comment (jerrypeng) [reply heron/downloaders/src/java/com/twitter/heron/downloader/FileDownloader.java]: sure thing

File: heron/config/src/yaml/conf/nomad/statemgr.yaml (hunk 9 above)
Comment (jerrypeng) [reply heron/config/src/yaml/conf/nomad/statemgr.yaml]: 2181 is the default port number for zookeeper.  If I was to run whole setup (scheduler, topology, and zookeeper) on a single node, this is the default I would use.  The cli and all executors will be able to contact the local zookeeper on 127.0.0.1:2181

File: heron/schedulers/tests/java/BUILD (hunk 14 above)
Comment (jerrypeng) [reply heron/schedulers/tests/java/BUILD]: will fix

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java (hunk 12 above)
Comment (jerrypeng) [reply heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: will fix

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java (hunk 13 above)
Comment (jerrypeng) [reply heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: will fix

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java (hunk 10 above)
Comment (jerrypeng) [reply heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: will fix

File: heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java (hunk 11 above)
Comment (jerrypeng) [reply heron/schedulers/tests/java/com/twitter/heron/scheduler/nomad/NomadSchedulerTest.java]: will fix

**Comment by jerrypeng** [comment]
@nwangtw thanks for the additional review!  I have addressed your comments

File: heron/config/src/yaml/conf/nomad/statemgr.yaml (hunk 9 above)
Comment (nwangtw) [reply heron/config/src/yaml/conf/nomad/statemgr.yaml]: got it. thx.

File: heron/config/src/yaml/BUILD (hunk 2 above)
Comment (nwangtw) [reply heron/config/src/yaml/BUILD]: kk.

File: heron/executor/src/python/heron_executor.py [hunk 15]
```diff
@@ -277,6 +280,7 @@ def parse_args(args):
     parser.add_argument("--topology-defn-file", required=True)
     parser.add_argument("--state-manager-connection", required=True)
     parser.add_argument("--state-manager-root", required=True)
+    parser.add_argument("--state-manager-config-file", required=True)
```
Comment (huijunw) [inline heron/executor/src/python/heron_executor.py]: config/yaml/heron.aurora needs update to add this param.

Comment (nwangtw) [reply heron/executor/src/python/heron_executor.py]: :( required. Is it possible to be optional?

Comment (nwangtw) [reply heron/executor/src/python/heron_executor.py]: I don't see any scheduler passes this flag. Why it is not causing errors?

Comment (jerrypeng) [reply heron/executor/src/python/heron_executor.py]: it is added in SchedulerUtils that every scheduler calls except aurora I believe

Comment (nwangtw) [reply heron/executor/src/python/heron_executor.py]: I see. That could be the case.

Another note is that it seems all the other schedulers have statemanager configs explicitly in the following file. It may make sense to have this new config as well, especially since it is a required field.

heron/config/src/yaml/conf/kubernetes/statemgr.yaml
heron/config/src/yaml/conf/marathon/statemgr.yaml
heron/config/src/yaml/conf/yarn/statemgr.yaml
and so on.

Comment (nwangtw) [reply heron/executor/src/python/heron_executor.py]: Never mind. I see what you mean. You want to override the config file itself.

Comment (nwangtw) [reply heron/executor/src/python/heron_executor.py]: @jerrypeng  Cant add you as reviewer. :(  Here is the PR for aurora when you get a minute. I am testing the change internally.

**Comment by dadgar** [comment]
@jerrypeng I am curious why not have heron submit a job to a running Nomad cluster?

**Comment by jerrypeng** [comment]
@dadgar heron can submit to an existing nomad scheduler, which is this PR focuses.  This PR allows users to submit heron jobs to existing nomad clusters.  In another PR, after this one was merged, we introduced a standalone cluster/distributed mode for heron using Nomad as the scheduler.  Before this mode was introduced, users would have to had setup Aurora, YARN, Kubernetes, etc. to run a Heron job using many nodes if they didn't already have a cluster of such available to them.  This was a high barrier for entry to users since many of these schedulers are not easy to setup, if not impossible, if you do not have sudo access.  Since Nomad is so simple to install and start up, we decided to use Nomad as the de facto scheduler/cluster manager for Heron standalone cluster mode.  The code to for setting up a standalone cluster is mainly scripts to start a nomad cluster and copying and templating files.  Like I said before, if a user already has Nomad cluster, they can absolutely use their existing deployment to run Heron jobs.

**Comment by jerrypeng** [comment]
@dadgar here is some instructions on how to start a standalone cluster via nomad:

https://twitter.github.io/heron/docs/operators/deployment/schedulers/standalone/

We still have to write some docs on how to deploy to existing nomad cluster

**Comment by dadgar** [comment]
@jerrypeng Ah thanks for the great answer! I saw the PR and then looked at the docs so I thought this was implementing the standalone mode! So there are just missing docs for running against an existing Nomad cluster.

That is awesome that you all chose to run the standalone cluster mode on Nomad!

**Comment by jerrypeng** [comment]
@dadgar yup!

@lucperkins if you have some time let's write some docs on how to deploy to an existing nomad cluster and also a blog to announce nomad support and standalone cluster mode.
